### PR TITLE
Make undo-tree an optional dependency

### PR DESCRIPTION
**Pull request by wasamasa**
I'd be thrilled if people tested this. Maybe even if a starter kit like Doom or Spacemacs enabled it. CC @hlissner @syl20bnr

### CODE REVIEW

**Comment by fbergroth** [comment]
Using `undo-redo`, I get an error when `evil-redo` calls `(undo-redo nil)`, i.e. when `count` is `nil`. It seems to work if I provide a digit argument.

**Comment by wasamasa** [comment]
@fbergroth Thanks, I've fixed this now.

**Comment by leungbk** [comment]
Thanks for doing this. It's working well for me.

**Comment by leungbk** [comment]
Since `goto-chg.el` is still a dependency, and since `undo-tree` is a dependency of `goto-chg.el`, `evil` still winds up depending on `undo-tree`. From a quick skim of `undo-tree`'s code, it doesn't look like the advice and hooks it sets are particularly intrusive when `global-undo-tree-mode` is turned off, though.

**Comment by wasamasa** [comment]
Hm, that's a good point. The code of goto-chg.el does have a soft dependency on undo-tree.el, but it still does declare a hard dependency on it and errors out if it cannot be loaded up successfully.

**Comment by TheBB** [comment]
Seems to work well for me, using the native Emacs (28+) functions.

File: evil-commands.el [hunk 1]
```diff
@@ -1623,12 +1623,12 @@ of the block."
 
 (evil-define-command evil-undo (count)
   "Undo COUNT changes in buffer using `evil-undo-function'."
-  (interactive "<c>")
+  (interactive "p")
```
Comment (fbergroth) [inline evil-commands.el:1626]: Perhaps `(interactive "*p")` (error if buffer is read-only)?

**Comment by wasamasa** [comment]
I've found that setting this to undo-tree only works if you've made sure to evaluate `(global-undo-tree-mode)` first. For some reason I haven't been successful with enabling it automatically for users, but eh, I can live with that.

File: evil-commands.el (hunk 1 above)
Comment (wasamasa) [reply evil-commands.el:1626]: Hm, it doesn't seem to ever hit that case sinde read-only buffers tend to not have any undo data in the first place or allowing you to edit them to create it.

Comment (fbergroth) [reply evil-commands.el:1626]: Ah ok. A user could potentially enable `read-only-mode` after having made some changes though. I just noticed that `undo-redo` used `"*p"`.

**Comment by TheBB** [comment]
Might be worth pointing that out somewhere, like the docstring of the *evil-undo-system* variable at least.

**Comment by wasamasa** [comment]
Alright, I've found out my issue and pushed a fix. Now you'll get a helpful error message if undo-tree-mode has not been enabled.

File: evil-commands.el (hunk 1 above)
Comment (wasamasa) [reply evil-commands.el]: Fair enough, fixed.

**Review by TheBB — APPROVED** [review]

**Comment by wasamasa** [comment]
Before merging this I'd like to try removing the hard dependency of goto-chg.el on undo-tree.el, otherwise the changes to the README claiming that Evil no longer requires undo-tree.el aren't true.

**Review by fbergroth — APPROVED** [review]

**Comment by wasamasa** [comment]
I went through the issue tracker and noticed that this PR might end up closing the following issues as well:

- https://github.com/emacs-evil/evil/issues/1275 (undo-tree is no longer vendored)
- https://github.com/emacs-evil/evil/issues/1144 (undo-tree is no longer the default)
- https://github.com/emacs-evil/evil/issues/1074 (the issue that prompted this PR)
- https://github.com/emacs-evil/evil/issues/601 (specific to undo-tree, cosmetic)
- https://github.com/emacs-evil/evil/issues/522 (specific to undo-tree)
- https://github.com/emacs-evil/evil/issues/418 (specific to undo-tree, resolved by using undo-fu)

**Comment by TheBB** [comment]
Good work. Thanks!

**Comment by JacksonGariety** [comment]
I now get the error `Customize 'evil-undo-system' for redo functionality.` when redoing in `evil-mode`. Is this expected to break? What's the suggested way to get redo now?

**Comment by leungbk** [comment]
> I now get the error `Customize 'evil-undo-system' for redo functionality.` when redoing in `evil-mode`. Is this expected to break? What's the suggested way to get redo now?

`(setq evil-undo-system PREFERRED-UNDO-SYSTEM)`

The options are  `'undo-tree` (extension), `'undo-fu` (extension),  and (Emacs 28 only) `'undo-redo`.

**Comment by TheBB** [comment]
Yeah it's expected. Do as the message says and customize that variable.

**Pull request merged**